### PR TITLE
change apply_rotary_pos_emb of Glmmodel for GLM-Edge Series model

### PR DESCRIPTION
**Pull request by zRzRzRzRzRzRzR**
# What does this PR do?

This PR is to allow the new version of the GLM-4 model to use different rotary_pos_emb.
I am still researching how to modify modular_glm.py so that model_glm.py can automatically generate an additional parameter called apply_rotary_pos_emb.

## Who can review?

This PR may [Cyrilvallez](https://github.com/Cyrilvallez) to help.

### CODE REVIEW

**Comment by zRzRzRzRzRzRzR** [comment]
Now this implementation is compatible with both GLM-Edge and GLM-4 models, @cyril-k, I would like to know how to modify modular_glm.py to achieve automatic updates, because some parts of the implementation in modeling_glm.py need to add new parameters to work properly.

**Review by Cyrilvallez — COMMENTED** [review]
Hey! Sorry for the delay, as I've said we were all in Martinique for our offsite the past week! 
You can check my comments, but unless I'm very much mistaken or missing something, most of the changes you propose are no-ops, and you only need to change `q, q_pass`/`k, k_pass` in `apply_rptary_pos_emb`
BTW, you tagged the wrong Cyril in the PR 🤣

File: src/transformers/models/glm/modeling_glm.py [hunk 1]
```diff
@@ -205,6 +181,7 @@ def __init__(self, config: GlmConfig, layer_idx: Optional[int] = None):
         self.num_key_value_groups = self.num_heads // self.num_key_value_heads
         self.is_causal = True
         self.scaling = 1 / math.sqrt(self.head_dim)
+        self.rotary_percent = config.rotary_percent if hasattr(config, "rotary_percent") else 0.5
```
[the remark this answers is not in the document]
Comment (Cyrilvallez) [reply src/transformers/models/glm/modeling_glm.py:184]: And let's use the name `partial_rotary_factor` instead of `rotary_percent`, for library consistency

File: src/transformers/models/glm/modeling_glm.py [hunk 2]
```diff
@@ -142,38 +132,24 @@ def rotate_half(x):
...
+    rotary_percent is for glm-4-9b(0.5) or glm-edge(1.0)
     """
     cos = cos.unsqueeze(unsqueeze_dim)
     sin = sin.unsqueeze(unsqueeze_dim)
 
     # Interleave them instead of usual shape
-    cos = cos[..., : cos.shape[-1] // 2].repeat_interleave(2, dim=-1)
-    sin = sin[..., : sin.shape[-1] // 2].repeat_interleave(2, dim=-1)
+    cos = cos[..., : int(cos.shape[-1] * rotary_percent)].repeat_interleave(2, dim=-1)
+    sin = sin[..., : int(sin.shape[-1] * rotary_percent)].repeat_interleave(2, dim=-1)
```
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/modeling_glm.py:145]: Because there are two different models, 0.5 and 1. In the config, the glm-edge series model needs to be set to 1.
https://huggingface.co/ZP2HF/glm-edge-4b-chat/blob/6a5e92d0092bba5f94abd471720238b6dda8f9de/config.json#L11
Here I have made annotations.

File: src/transformers/models/glm/modeling_glm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zRzRzRzRzRzRzR) [reply src/transformers/models/glm/modeling_glm.py:184]: Yes, I can make modifications to this part.

**Comment by zRzRzRzRzRzRzR** [comment]
I referred to your plan and modified it to look like this. I believe the mathematical logic of this implementation is equivalent.

File: src/transformers/models/glm/modeling_glm.py [hunk 3]
```diff
@@ -142,44 +123,28 @@ def rotate_half(x):
...
+    rotary_dim = int(q.shape[-1] * partial_rotary_factor)
+    q, q_pass = q[..., :rotary_dim], q[..., rotary_dim:]
+    k, k_pass = k[..., :rotary_dim], k[..., rotary_dim:]
 
-    # Apply rotary embeddings on the first half
-    q_embed = (q * cos) + (rotate_half(q) * sin)
-    k_embed = (k * cos) + (rotate_half(k) * sin)
+    # Apply rotary embeddings to the rotary portion
+    q = (q * cos[..., :rotary_dim]) + (rotate_half(q) * sin[..., :rotary_dim])
+    k = (k * cos[..., :rotary_dim]) + (rotate_half(k) * sin[..., :rotary_dim])
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py]: Here you don't need to modify anything, you are basically slicing up to the full length which is useless

File: src/transformers/models/glm/modeling_glm.py [hunk 4]
```diff
@@ -698,14 +690,16 @@ def __init__(self, config: GlmConfig):
...
         self.layers = nn.ModuleList(
             [GlmDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
         )
         self.norm = GlmRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.partial_rotary_factor = config.partial_rotary_factor
         self.rotary_emb = GlmRotaryEmbedding(
-            dim=config.head_dim // 2, max_position_embeddings=config.max_position_embeddings, base=config.rope_theta
+            dim=config.head_dim * self.partial_rotary_factor,
+            max_position_embeddings=config.max_position_embeddings,
+            base=config.rope_theta,
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py:718]: Yes, that's indeed the only change that should be applied to the RotaryEmbedding 🤗

**Review by Cyrilvallez — COMMENTED** [review]
It's getting better! Still some issue in the rotary though I think
Then, once we agree on the changes, you'll need to apply the changes in `modular` instead of `modeling` 🤗

**Comment by zRzRzRzRzRzRzR** [comment]
I would like to know if there are any improvements needed for this version, and also, I would like to know if @Cyrilvallez could guide me on how to modify modular_glm.py to make good changes.
```
query_states, key_states = apply_rotary_pos_emb(
            query_states, key_states, cos, sin, partial_rotary_factor=self.partial_rotary_factor
        )
```

Let it automatically generate to modeling_glm.py.

**Comment by ArthurZucker** [comment]
cc @Cyrilvallez but I think it would help to remove unrelated changes! 🤗

File: src/transformers/models/glm/modeling_glm.py [hunk 5]
```diff
@@ -169,11 +170,12 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
     cos = cos[..., : cos.shape[-1] // 2].repeat_interleave(2, dim=-1)
     sin = sin[..., : sin.shape[-1] // 2].repeat_interleave(2, dim=-1)
 
-    # Keep half for later concatenation
-    q, q_pass = q[..., : q.shape[-1] // 2], q[..., q.shape[-1] // 2 :]
-    k, k_pass = k[..., : k.shape[-1] // 2], k[..., k.shape[-1] // 2 :]
+    # Keep half or full tensor for later concatenation
+    rotary_dim = int(q.shape[-1] * partial_rotary_factor)
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py:174]:
```suggestion
    # Keep half or full tensor for later concatenation
    rotary_dim = cos.shape[-1]
```
We actually don't need to pass the `rotary_factor` as an argument to the function!

File: src/transformers/models/glm/modeling_glm.py [hunk 6]
```diff
@@ -142,7 +142,7 @@ def rotate_half(x):
     return torch.stack((-x2, x1), dim=-1).flatten(-2)
 
 
-def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1, partial_rotary_factor=0.5):
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py:145]:
```suggestion
def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
We actually don't need to pass the rotary_factor as an argument to the function! See next comment, that way we don't even have to modify the modular file for the Attentions!

File: src/transformers/models/glm/modeling_glm.py [hunk 7]
```diff
@@ -240,7 +243,9 @@ def forward(
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         cos, sin = position_embeddings
-        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
+        query_states, key_states = apply_rotary_pos_emb(
+            query_states, key_states, cos, sin, partial_rotary_factor=self.partial_rotary_factor
+        )
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py:248]: No need to pass the extra arg! See above

File: src/transformers/models/glm/modeling_glm.py [hunk 8]
```diff
@@ -320,7 +325,9 @@ def forward(
...
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py:330]: Same

File: src/transformers/models/glm/modeling_glm.py [hunk 9]
```diff
@@ -432,7 +439,9 @@ def forward(
...
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modeling_glm.py:444]: Same

File: src/transformers/models/glm/modular_glm.py [hunk 10]
```diff
@@ -68,7 +68,7 @@ def rotate_half(x):
     return torch.stack((-x2, x1), dim=-1).flatten(-2)
 
 
-def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1, partial_rotary_factor=0.5):
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modular_glm.py:71]: No need to pass the extra arg! See above

File: src/transformers/models/glm/modular_glm.py [hunk 11]
```diff
@@ -85,6 +85,7 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
             k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
             cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
             the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
+        partial_rotary_factor (`float`, *optional*, defaults to 0.5): The factor by which the rotary embedding.
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modular_glm.py:88]: Same, can be removed

File: src/transformers/models/glm/modular_glm.py [hunk 12]
```diff
@@ -95,11 +96,12 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
     cos = cos[..., : cos.shape[-1] // 2].repeat_interleave(2, dim=-1)
     sin = sin[..., : sin.shape[-1] // 2].repeat_interleave(2, dim=-1)
 
-    # Keep half for later concatenation
-    q, q_pass = q[..., : q.shape[-1] // 2], q[..., q.shape[-1] // 2 :]
-    k, k_pass = k[..., : k.shape[-1] // 2], k[..., k.shape[-1] // 2 :]
+    # Keep half or full tensor for later concatenation
+    rotary_dim = int(q.shape[-1] * partial_rotary_factor)
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modular_glm.py:100]:
```suggestion
    # Keep half or full tensor for later concatenation
    rotary_dim = cos.shape[-1]
```
Same as above

**Review by Cyrilvallez — COMMENTED** [review]
Nice! That's it! 🤗
Final comment is that we don't even need to change the signature of `apply_rotary_pos_embed` as we can retrieve the `rotary_dim` from `cos` and `sin`, I forgot before sorry! That way, we don't even have to modify the attention implementation, which is a big win for the modular!

Also, please remove the unrelated notebook change you added (I assume as a mistake) 🤗

File: src/transformers/models/glm/modular_glm.py [hunk 13]
```diff
@@ -151,8 +154,11 @@ def __init__(self, config: GlmConfig):
             [GlmDecoderLayer(config, layer_idx) for layer_idx in range(config.num_hidden_layers)]
         )
         self.norm = GlmRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.partial_rotary_factor = config.partial_rotary_factor
         self.rotary_emb = GlmRotaryEmbedding(
-            dim=config.head_dim // 2, max_position_embeddings=config.max_position_embeddings, base=config.rope_theta
+            dim=config.head_dim * config.partial_rotary_factor,
```
Comment (Cyrilvallez) [inline src/transformers/models/glm/modular_glm.py:159]:
```suggestion
            dim=int(config.head_dim * config.partial_rotary_factor),
```
You need `int` here as well

**Review by Cyrilvallez — COMMENTED** [review]
Sorry, forgot a comment, you need to ensure the dim in an integer as well.

To automatically generate the modeling file from the modular, you can run 
```sh
python utils/modular_model_converter.py --files_to_parse src/transformers/models/glm/modular_glm.py
```
from the root of the `transformers` repo 🤗

**Comment by zRzRzRzRzRzRzR** [comment]
This modification should meet the requirements, and I have tried to remove all unnecessary code. The remaining code is all the code that will be used.

**Review by Cyrilvallez — APPROVED** [review]
All good, thanks for iterating!

**Review by ArthurZucker — COMMENTED** [review]
Thanks, small convention for q_rot, q_pass, and a nit! We can merge afterwards!

**Comment by Cyrilvallez** [comment]
All good, merging!
